enhancements/sig-architecture/227-manifestworkreplicaset-rollout-plugin/README.md

```diff
@@ -0,0 +1,342 @@
+          MWRSController -->HelloPlacement[Placement: HelloPlacement]
+        end
+
+        subgraph NamespaceCluster1[Namespace: cluster1]
+          MWRSController -->|Apply/Delete| MW1[ManifestWork: HelloRollout]
+        end
+
+        subgraph NamespaceCluster2[Namespace: cluster2]
+          MWRSController -->|Apply/Delete| MW2[ManifestWork: HelloRollout]
+        end
+    end
+    
+    MeshService[Istio ControlPlane]
+
+    Plugin -->|Update traffic weights| MeshService
+```
+
+To enable the user stories described above, the following custom hooks will be required:
+
+| Hook                     | Description                                                                                 |
+| ---------------------------- | ----------------------------------------------------------------------------------------------- |
+| `BeginRollout`               | Called before rollout starts. Used to prepare metadata, preconditions, or environment setup. |
+| `ProgressRollout`            | Called every MWRS reconciler is triggered during rollout. |
+| `ValidateRolloutCompletion`  | Called when cluster rollout finishes. Plugin determines success, failure, or in-progress state. |
+| `BeginRollback`              | Called before rollback starts to prepare environment. |
+| `ProgressRollback`           | Called every MWRS reconciler is triggered during rollback. |
+| `ValidateRollbackCompletion` | Called after rollback completion for verification. |
+| `MutateManifestWork`         | Called before creating or updating ManifestWork to modify manifest templates dynamically. |
+
+The design extends the work-controller with these custom hooks, which are implemented using a gRPC protocol between the controller and a plugin sidecar.
+
+### Plugin protocol
+
+The plugin will be implemented as a sidecar container running alongside the work controller in hub cluster. We chose a sidecar model to simplify the plugin onboarding process. gRPC will be used for communication. The plugin will run as a gRPC server, and the work controller will act as the gRPC client.
+
+#### Plugin initialization
+
+When the work controller at hub starts, it attempts to call the `Initialize()` gRPC endpoint on the plugin server. If successful, the plugin returns its supported capabilities (e.g., Rollout, Rollback, MutateManifestWork). The reconciler will set the MWRS resource status with a `PluginLoaded` condition. If no plugin is configured, this condition will not be present.
+
+```yaml
+apiVersion: work.open-cluster-management.io/v1alpha1
+kind: ManifestWorkReplicaSet
+metadata:
+  name: hello-rollout
+  namespace: default
+spec:
+  ...
+status:
+  conditions:
+    - lastTransitionTime: "2025-10-09T04:40:41Z"
+      message: "Plugin is loaded successfully. Supported capability: Rollout, Rollback, mutateManifestWork."
+      observedGeneration: 1
+      reason: PluginInitialized
+      status: "True"
+      type: PluginLoaded
```

Comment:
I wonder if it is good idea to expose the plugin status or not... simple logging in work controller might be enough. I would like to get the feedback.

Reply:
this status might not be needed, but we would need to surface the message on mwrs API when calling grpc API fails.